File: v2v_cooperative_pkg/v2v_cooperative_pkg/reverse_path_planner_node.py
#!/usr/bin/env python3
"""
reverse_path_planner_node
──────────────────────────
A* 기반 후진 회피 경로 생성 노드.

현재 위치에서 reverse_goal까지 점유 격자 지도(OccupancyGrid) 위에서
A* 탐색으로 전역 경로를 생성한다.
경로 방향은 후진이므로 waypoint 순서가 역방향이 된다.

입력 토픽:
  /map            nav_msgs/OccupancyGrid  (slam_toolbox)
  /odom           nav_msgs/Odometry
  /reverse_goal   geometry_msgs/PoseStamped  (spatial_memory_node)
  /vehicle_state  std_msgs/String

출력 토픽:
  /reverse_path   nav_msgs/Path
  /reverse_path_ready Bool
"""

import heapq
import math
from typing import Dict, List, Optional, Tuple

import numpy as np
import rclpy
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import OccupancyGrid, Odometry, Path
from rclpy.node import Node
from std_msgs.msg import Bool, String


def quat_to_yaw(q) -> float:
    siny = 2.0 * (q.w * q.z + q.x * q.y)
    cosy = 1.0 - 2.0 * (q.y * q.y + q.z * q.z)
    return math.atan2(siny, cosy)


# ── A* 구현 ───────────────────────────────────────────────────────────────
def _heuristic(a: Tuple[int, int], b: Tuple[int, int]) -> float:
    return math.hypot(a[0] - b[0], a[1] - b[1])


def astar(
    grid: np.ndarray,          # (rows, cols), True=장애물
    start: Tuple[int, int],
    goal: Tuple[int, int],
    robot_radius_cells: int = 2,
) -> Optional[List[Tuple[int, int]]]:
    """
    A* 경로 탐색. 반환값: 셀 좌표 리스트(start→goal) 또는 None.
    robot_radius_cells: 로봇 크기 고려한 팽창 반경.
    """
    rows, cols = grid.shape

    # 장애물 팽창 (Minkowski sum 근사)
    from scipy.ndimage import binary_dilation  # type: ignore
    structure = np.ones((2 * robot_radius_cells + 1,) * 2, dtype=bool)
    inflated = binary_dilation(grid, structure=structure)

    def is_free(r: int, c: int) -> bool:
        return 0 <= r < rows and 0 <= c < cols and not inflated[r, c]

    if not is_free(*start) or not is_free(*goal):
        return None

    open_heap: List[Tuple[float, int, int]] = []
    heapq.heappush(open_heap, (0.0, start[0], start[1]))

    came_from: Dict[Tuple[int, int], Optional[Tuple[int, int]]] = {start: None}
    g_score: Dict[Tuple[int, int], float] = {start: 0.0}

    neighbors_8 = [(-1, 0), (1, 0), (0, -1), (0, 1),
                   (-1, -1), (-1, 1), (1, -1), (1, 1)]

    while open_heap:
        _, r, c = heapq.heappop(open_heap)
        cur = (r, c)
        if cur == goal:
            # 경로 재구성
            path: List[Tuple[int, int]] = []
            node: Optional[Tuple[int, int]] = goal
            while node is not None:
                path.append(node)
                node = came_from[node]
            path.reverse()
            return path

        for dr, dc in neighbors_8:
            nr, nc = r + dr, c + dc
            nb = (nr, nc)
            if not is_free(nr, nc):
                continue
            move_cost = math.hypot(dr, dc)
            tentative_g = g_score[cur] + move_cost
            if nb not in g_score or tentative_g < g_score[nb]:
                g_score[nb] = tentative_g
                f = tentative_g + _heuristic(nb, goal)
                heapq.heappush(open_heap, (f, nr, nc))
                came_from[nb] = cur

    return None  # 경로 없음


class ReversePathPlannerNode(Node):
    def __init__(self) -> None:
        super().__init__('reverse_path_planner_node')

        self.declare_parameter('planner_hz',         2.0)
        self.declare_parameter('robot_width_m',      0.19)
        self.declare_parameter('path_simplify_dist', 0.1)  # 경로 단순화 간격 (m)
        self.declare_parameter('occ_threshold',      50)   # 점유 격자 점유 임계값

        self.planner_hz        = float(self.get_parameter('planner_hz').value)
        self.robot_width_m     = float(self.get_parameter('robot_width_m').value)
        self.path_simplify_dist = float(self.get_parameter('path_simplify_dist').value)
        self.occ_threshold     = int(self.get_parameter('occ_threshold').value)

        # 내부 상태
        self.map_msg:     Optional[OccupancyGrid] = None
        self.odom_msg:    Optional[Odometry]      = None
        self.goal_msg:    Optional[PoseStamped]   = None
        self.vehicle_state: str = 'NORMAL_CENTER_DRIVE'

        self.last_path:   Optional[Path] = None
        self.path_ready:  bool = False

        # 구독
        self.create_subscription(OccupancyGrid, '/map',           self._cb_map,   1)
        self.create_subscription(Odometry,      '/odom',          self._cb_odom,  10)
        self.create_subscription(PoseStamped,   '/reverse_goal',  self._cb_goal,  10)
        self.create_subscription(String,        '/vehicle_state', self._cb_state, 10)

        # 발행
        self.pub_path  = self.create_publisher(Path, '/reverse_path',       1)
        self.pub_ready = self.create_publisher(Bool, '/reverse_path_ready', 10)

        dt = 1.0 / self.planner_hz if self.planner_hz > 0 else 0.5
        self.create_timer(dt, self._step)
        self.get_logger().info('reverse_path_planner_node 시작')

    # ── 콜백 ─────────────────────────────────────────────────────────────
    def _cb_map(self, msg: OccupancyGrid) -> None:
        self.map_msg = msg

    def _cb_odom(self, msg: Odometry) -> None:
        self.odom_msg = msg

    def _cb_goal(self, msg: PoseStamped) -> None:
        self.goal_msg = msg

    def _cb_state(self, msg: String) -> None:
        prev = self.vehicle_state
        self.vehicle_state = str(msg.data).strip()
        if prev != 'REVERSE_EXECUTE' and self.vehicle_state == 'REVERSE_EXECUTE':
            # REVERSE_EXECUTE 진입 시 경로 재계산
            self._plan()

    # ── 좌표 변환 ─────────────────────────────────────────────────────────
    def _world_to_cell(self, wx: float, wy: float) -> Optional[Tuple[int, int]]:
        if self.map_msg is None:
            return None
        info = self.map_msg.info
        ox = info.origin.position.x
        oy = info.origin.position.y
        res = info.resolution
        col = int((wx - ox) / res)
        row = int((wy - oy) / res)
        if 0 <= row < info.height and 0 <= col < info.width:
            return row, col
        return None

    def _cell_to_world(self, row: int, col: int) -> Tuple[float, float]:
        info = self.map_msg.info
        ox = info.origin.position.x
        oy = info.origin.position.y
        res = info.resolution
        wx = ox + (col + 0.5) * res
        wy = oy + (row + 0.5) * res
        return wx, wy

    def _build_grid(self) -> Optional[np.ndarray]:
        if self.map_msg is None:
            return None
        info = self.map_msg.info
        data = np.array(self.map_msg.data, dtype=np.int8).reshape(info.height, info.width)
        # 점유(>=threshold) 또는 미지(-1)를 장애물로 처리
        grid = (data >= self.occ_threshold) | (data < 0)
        return grid

    # ── 경로 단순화 ──────────────────────────────────────────────────────
    def _simplify_path(self, pts: List[Tuple[float, float]]) -> List[Tuple[float, float]]:
        if len(pts) < 2:
            return pts
        result = [pts[0]]
        for p in pts[1:]:
            if math.hypot(p[0] - result[-1][0], p[1] - result[-1][1]) >= self.path_simplify_dist:
                result.append(p)
        if result[-1] != pts[-1]:
            result.append(pts[-1])
        return result

    # ── A* 경로 계획 ─────────────────────────────────────────────────────
    def _plan(self) -> None:
        if self.map_msg is None or self.odom_msg is None or self.goal_msg is None:
            self.path_ready = False
            return

        # 시작 = 현재 위치
        sx = float(self.odom_msg.pose.pose.position.x)
        sy = float(self.odom_msg.pose.pose.position.y)
        # 목표 = reverse_goal
        gx = float(self.goal_msg.pose.position.x)
        gy = float(self.goal_msg.pose.position.y)

        start_cell = self._world_to_cell(sx, sy)
        goal_cell  = self._world_to_cell(gx, gy)

        if start_cell is None or goal_cell is None:
            self.get_logger().warn('[Planner] start/goal 셀 변환 실패')
            self.path_ready = False
            return

        grid = self._build_grid()
        if grid is None:
            self.path_ready = False
            return

        res = self.map_msg.info.resolution
        robot_radius_cells = max(1, int(math.ceil(self.robot_width_m / 2.0 / res)))

        self.get_logger().info(
            f'[Planner] A* 탐색: ({start_cell}) → ({goal_cell})'
        )
        try:
            cell_path = astar(grid, start_cell, goal_cell, robot_radius_cells)
        except Exception as e:
            self.get_logger().error(f'[Planner] A* 오류: {e}')
            self.path_ready = False
            return

        if cell_path is None:
            self.get_logger().warn('[Planner] 경로 없음')
            self.path_ready = False
            return

        # 셀 → 월드 좌표
        world_pts = [self._cell_to_world(r, c) for r, c in cell_path]
        world_pts = self._simplify_path(world_pts)

        # nav_msgs/Path 생성
        path = Path()
        path.header.stamp = self.get_clock().now().to_msg()
        path.header.frame_id = 'map'
        for wx, wy in world_pts:
            ps = PoseStamped()
            ps.header = path.header
            ps.pose.position.x = wx
            ps.pose.position.y = wy
            ps.pose.position.z = 0.0
            ps.pose.orientation.w = 1.0
            path.poses.append(ps)

        self.last_path = path
        self.path_ready = True
        self.get_logger().info(f'[Planner] 경로 생성 완료: {len(world_pts)} 웨이포인트')

    # ── 메인 스텝 ─────────────────────────────────────────────────────────
    def _step(self) -> None:
        # REVERSE_EXECUTE 상태에서만 경로 발행 (매 스텝 재계획은 하지 않음)
        ready_msg = Bool(); ready_msg.data = self.path_ready
        self.pub_ready.publish(ready_msg)

        if self.path_ready and self.last_path is not None:
            self.last_path.header.stamp = self.get_clock().now().to_msg()
            self.pub_path.publish(self.last_path)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = ReversePathPlannerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: v2v_cpp_nodes/src/reverse_path_planner_node.cpp
/**
 * reverse_path_planner_node (C++)
 * ─────────────────────────────────
 * A* 기반 후진 회피 경로 생성.
 *
 * 점유 격자 지도(OccupancyGrid) 위에서 현재 위치 → reverse_goal 까지
 * A* 탐색으로 경로를 생성한다.
 */

#include <cmath>
#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>
#include <string>
#include <functional>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/bool.hpp"
#include "std_msgs/msg/string.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "nav_msgs/msg/occupancy_grid.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "nav_msgs/msg/path.hpp"

// ── A* 구현 ─────────────────────────────────────────────────────────────
struct Cell {
    int r, c;
    bool operator==(const Cell &o) const { return r == o.r && c == o.c; }
};

struct CellHash {
    size_t operator()(const Cell &c) const {
        return std::hash<int>()(c.r) ^ (std::hash<int>()(c.c) << 16);
    }
};

static double heuristic(Cell a, Cell b) {
    return std::hypot(a.r - b.r, a.c - b.c);
}

struct AStarEntry {
    double f;
    Cell cell;
    bool operator>(const AStarEntry &o) const { return f > o.f; }
};

static std::vector<Cell> astar(
    const std::vector<bool> &occupied,
    int rows, int cols,
    Cell start, Cell goal,
    int robot_radius_cells)
{
    // ── 장애물 팽창 ─────────────────────────────────────────────────
    std::vector<bool> inflated(rows * cols, false);
    for (int r = 0; r < rows; ++r) {
        for (int c = 0; c < cols; ++c) {
            if (!occupied[r * cols + c]) continue;
            for (int dr = -robot_radius_cells; dr <= robot_radius_cells; ++dr) {
                for (int dc = -robot_radius_cells; dc <= robot_radius_cells; ++dc) {
                    int nr = r + dr, nc = c + dc;
                    if (nr >= 0 && nr < rows && nc >= 0 && nc < cols)
                        inflated[nr * cols + nc] = true;
                }
            }
        }
    }

    auto is_free = [&](int r, int c) -> bool {
        return r >= 0 && r < rows && c >= 0 && c < cols && !inflated[r * cols + c];
    };

    if (!is_free(start.r, start.c) || !is_free(goal.r, goal.c))
        return {};

    std::priority_queue<AStarEntry, std::vector<AStarEntry>, std::greater<>> open;
    std::unordered_map<Cell, Cell, CellHash> came_from;
    std::unordered_map<Cell, double, CellHash> g_score;

    open.push({0.0, start});
    g_score[start] = 0.0;
    came_from[start] = {-1, -1};  // sentinel

    static const int dx[] = {-1, 1, 0, 0, -1, -1, 1, 1};
    static const int dy[] = {0, 0, -1, 1, -1, 1, -1, 1};

    while (!open.empty()) {
        auto cur = open.top().cell;
        open.pop();

        if (cur == goal) {
            std::vector<Cell> path;
            Cell node = goal;
            while (!(node.r == -1 && node.c == -1)) {
                path.push_back(node);
                node = came_from[node];
            }
            std::reverse(path.begin(), path.end());
            return path;
        }

        for (int i = 0; i < 8; ++i) {
            Cell nb{cur.r + dx[i], cur.c + dy[i]};
            if (!is_free(nb.r, nb.c)) continue;
            double move_cost = std::hypot(dx[i], dy[i]);
            double tentative_g = g_score[cur] + move_cost;
            auto it = g_score.find(nb);
            if (it == g_score.end() || tentative_g < it->second) {
                g_score[nb] = tentative_g;
                double f = tentative_g + heuristic(nb, goal);
                open.push({f, nb});
                came_from[nb] = cur;
            }
        }
    }
    return {};  // 경로 없음
}


// ── 노드 ────────────────────────────────────────────────────────────────
static double quat_to_yaw(const geometry_msgs::msg::Quaternion &q) {
    double siny = 2.0 * (q.w * q.z + q.x * q.y);
    double cosy = 1.0 - 2.0 * (q.y * q.y + q.z * q.z);
    return std::atan2(siny, cosy);
}

class ReversePathPlannerNode : public rclcpp::Node {
public:
    ReversePathPlannerNode() : Node("reverse_path_planner_node") {
        this->declare_parameter("planner_hz",         2.0);
        this->declare_parameter("robot_width_m",      0.19);
        this->declare_parameter("path_simplify_dist", 0.1);
        this->declare_parameter("occ_threshold",      50);

        planner_hz_        = this->get_parameter("planner_hz").as_double();
        robot_width_m_     = this->get_parameter("robot_width_m").as_double();
        path_simplify_dist_= this->get_parameter("path_simplify_dist").as_double();
        occ_threshold_     = this->get_parameter("occ_threshold").as_int();

        sub_map_  = create_subscription<nav_msgs::msg::OccupancyGrid>(
            "/map", 1,
            [this](nav_msgs::msg::OccupancyGrid::SharedPtr m) {
                map_ = m;
                plan_if_reversing();
            });

        sub_odom_ = create_subscription<nav_msgs::msg::Odometry>(
            "/odom", 10,
            [this](nav_msgs::msg::Odometry::SharedPtr m) { odom_ = m; });

        sub_goal_ = create_subscription<geometry_msgs::msg::PoseStamped>(
            "/reverse_goal", 10,
            [this](geometry_msgs::msg::PoseStamped::SharedPtr m) {
                goal_ = m;
                path_ready_ = false;
                plan_if_reversing();
            });

        sub_state_ = create_subscription<std_msgs::msg::String>(
            "/vehicle_state", 10,
            [this](std_msgs::msg::String::SharedPtr m) {
                auto prev = vehicle_state_;
                vehicle_state_ = m->data;
                if (prev != "REVERSE_EXECUTE" && vehicle_state_ == "REVERSE_EXECUTE")
                    plan();
                if (vehicle_state_ != "REVERSE_EXECUTE")
                    path_ready_ = false;
            });

        pub_path_  = create_publisher<nav_msgs::msg::Path>("/reverse_path", 1);
        pub_ready_ = create_publisher<std_msgs::msg::Bool>("/reverse_path_ready", 10);

        double dt = (planner_hz_ > 0) ? (1.0 / planner_hz_) : 0.5;
        timer_ = create_wall_timer(
            std::chrono::duration<double>(dt),
            std::bind(&ReversePathPlannerNode::step, this));

        RCLCPP_INFO(get_logger(), "reverse_path_planner_node (C++) 시작");
    }

private:
    void plan_if_reversing() {
        if (vehicle_state_ == "REVERSE_EXECUTE")
            plan();
    }

    bool world_to_cell(double wx, double wy, int &row, int &col) const {
        if (!map_) return false;
        auto &info = map_->info;
        double ox = info.origin.position.x;
        double oy = info.origin.position.y;
        double res = info.resolution;
        if (res < 1e-6) return false;
        col = static_cast<int>((wx - ox) / res);
        row = static_cast<int>((wy - oy) / res);
        return row >= 0 && row < static_cast<int>(info.height) &&
               col >= 0 && col < static_cast<int>(info.width);
    }

    std::pair<double, double> cell_to_world(int row, int col) const {
        auto &info = map_->info;
        double ox = info.origin.position.x;
        double oy = info.origin.position.y;
        double res = info.resolution;
        return {ox + (col + 0.5) * res, oy + (row + 0.5) * res};
    }

    void plan() {
        if (!map_ || !odom_ || !goal_) {
            path_ready_ = false;
            return;
        }

        double sx = odom_->pose.pose.position.x;
        double sy = odom_->pose.pose.position.y;
        double gx = goal_->pose.position.x;
        double gy = goal_->pose.position.y;

        int sr, sc, gr, gc;
        if (!world_to_cell(sx, sy, sr, sc) || !world_to_cell(gx, gy, gr, gc)) {
            RCLCPP_WARN(get_logger(), "[Planner] 셀 변환 실패");
            path_ready_ = false;
            return;
        }

        auto &info = map_->info;
        int rows = info.height;
        int cols = info.width;

        // 점유 격자 생성
        std::vector<bool> occ(rows * cols);
        for (int i = 0; i < rows * cols; ++i) {
            int8_t v = map_->data[i];
            occ[i] = (v >= occ_threshold_) || (v < 0);
        }

        double res = info.resolution;
        int radius = std::max(1, static_cast<int>(std::ceil(robot_width_m_ / 2.0 / res)));

        RCLCPP_INFO(get_logger(), "[Planner] A*: (%d,%d) → (%d,%d)", sr, sc, gr, gc);

        auto cell_path = astar(occ, rows, cols, {sr, sc}, {gr, gc}, radius);
        if (cell_path.empty()) {
            RCLCPP_WARN(get_logger(), "[Planner] 경로 없음");
            path_ready_ = false;
            return;
        }

        // 경로 단순화 + Path 메시지 생성
        nav_msgs::msg::Path path;
        path.header.stamp = now();
        path.header.frame_id = "map";

        double last_wx = -1e9, last_wy = -1e9;
        for (auto &c : cell_path) {
            auto [wx, wy] = cell_to_world(c.r, c.c);
            if (std::hypot(wx - last_wx, wy - last_wy) < path_simplify_dist_)
                continue;
            geometry_msgs::msg::PoseStamped ps;
            ps.header = path.header;
            ps.pose.position.x = wx;
            ps.pose.position.y = wy;
            ps.pose.orientation.w = 1.0;
            path.poses.push_back(ps);
            last_wx = wx; last_wy = wy;
        }
        // 마지막 점 보장
        auto [fwx, fwy] = cell_to_world(cell_path.back().r, cell_path.back().c);
        if (std::hypot(fwx - last_wx, fwy - last_wy) > 0.01) {
            geometry_msgs::msg::PoseStamped ps;
            ps.header = path.header;
            ps.pose.position.x = fwx;
            ps.pose.position.y = fwy;
            ps.pose.orientation.w = 1.0;
            path.poses.push_back(ps);
        }

        last_path_ = path;
        path_ready_ = true;
        RCLCPP_INFO(get_logger(), "[Planner] 경로 생성: %zu 웨이포인트",
                     path.poses.size());
    }

    void step() {
        std_msgs::msg::Bool b;
        b.data = path_ready_;
        pub_ready_->publish(b);

        if (path_ready_) {
            last_path_.header.stamp = now();
            pub_path_->publish(last_path_);
        }
    }

    // ── 파라미터 ─────────────────────────────────────────────────
    double planner_hz_, robot_width_m_, path_simplify_dist_;
    int occ_threshold_;

    // ── 상태 ─────────────────────────────────────────────────────
    nav_msgs::msg::OccupancyGrid::SharedPtr map_;
    nav_msgs::msg::Odometry::SharedPtr odom_;
    geometry_msgs::msg::PoseStamped::SharedPtr goal_;
    std::string vehicle_state_{"NORMAL_CENTER_DRIVE"};
    nav_msgs::msg::Path last_path_;
    bool path_ready_{false};

    // ── ROS ───────────────────────────────────────────────────────
    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr sub_map_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_odom_;
    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr sub_goal_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_state_;
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr pub_path_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr pub_ready_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ReversePathPlannerNode>());
    rclcpp::shutdown();
    return 0;
}
